docs/manuscripts/cosmicqc_paper_v1/figure_5/nbconverted/5.calculate_mAP.py

```diff
@@ -0,0 +1,585 @@
+post_qc_df = pd.read_parquet(post_qc_file)
+post_qc_df = post_qc_df.drop(
+    columns=["broad_id", "pert_iname", "moa", "replicate_name"]
+)
+post_qc_df["Metadata_sc_count_failed_qc"] = post_qc_df[
+    "Metadata_sc_count_failed_qc"
+].fillna(0)
+
+# Align "Metadata_sc_count_failed_qc" from post-QC df to pre-QC df using
+# Metadata_Plate and Metadata_Well
+pre_qc_df = pre_qc_df.merge(
+    post_qc_df[["Metadata_Plate", "Metadata_Well", "Metadata_sc_count_failed_qc"]],
+    on=["Metadata_Plate", "Metadata_Well"],
+    how="left",
+    suffixes=("", "_postqc"),
+)
+pre_qc_df["Metadata_sc_count_failed_qc"] = pre_qc_df[
+    "Metadata_sc_count_failed_qc"
+].fillna(0)
+
+
+# In[4]:
+
+
+# Compute percentage of failed QC single-cell across all plates in post_qc_df
+total_failed = post_qc_df["Metadata_sc_count_failed_qc"].sum()
+total_cells = post_qc_df["Metadata_sc_count"].sum()
+failed_qc_percentage = (total_failed / total_cells) * 100
+print(
+    f"Percentage of failed QC single-cell across all plates: {failed_qc_percentage:.2f}%"
+)
+
+
+# In[5]:
+
+
+# Calculate proportion of failed cells for each well
+post_qc_df["failed_proportion"] = post_qc_df["Metadata_sc_count_failed_qc"] / (
+    post_qc_df["Metadata_sc_count_failed_qc"]
+    + post_qc_df["Metadata_sc_count_passed_qc"]
+)
+
+# Compare failed proportions between compounds and DMSO
+compound_failed = post_qc_df.loc[
+    post_qc_df["Metadata_broad_sample"] != "DMSO", "failed_proportion"
+]
+dmso_failed = post_qc_df.loc[
+    post_qc_df["Metadata_broad_sample"] == "DMSO", "failed_proportion"
+]
+
+# Print mean failed proportions
+print(f"Mean failed proportion (compounds): {compound_failed.mean():.4f}")
+print(f"Mean failed proportion (DMSO): {dmso_failed.mean():.4f}")
+
+
+# ## Assign reference index
+
+# In[6]:
+
+
+reference_col = "Metadata_reference_index"
+
+pre_qc_df_activity = assign_reference_index(
+    pre_qc_df,
+    "Metadata_broad_sample == 'DMSO'",  # condition to get reference profiles (neg controls)
+    reference_col=reference_col,
+    default_value=-1,
+)
+
+post_qc_df_activity = assign_reference_index(
+    post_qc_df,
+    "Metadata_broad_sample == 'DMSO'",  # condition to get reference profiles (neg controls)
+    reference_col=reference_col,
+    default_value=-1,
+)
+
+
+# ## Set positive and negative pairs for compounds
+
+# In[7]:
+
+
+# positive pairs are replicates of the same treatment
+pos_sameby = ["Metadata_broad_sample", "Metadata_dose_recode", reference_col]
+pos_diffby = ["Metadata_Plate"]
+
+# negative pairs are replicates of different treatments
+neg_sameby = []  # set plate if you don't want to compare controls across all plates
+neg_diffby = ["Metadata_broad_sample", reference_col]
+
+
+# ## Calculate mAP scores per treatment compared to controls for both pre- and post-QC
+
+# ### Pre-QC dataframe
+
+# In[8]:
+
+
+list_of_dfs_map_preQC = []
+for treatment in pre_qc_df_activity["Metadata_broad_sample"].unique():
```

Comment:
These for loops similar, is there a way to just use one for loop for both pre/post qc profiles?